internal/log_analysis/log_processor/processor/rescale.go

```
@@ -49,49 +53,67 @@ var (
 
 // scalingDecisions makes periodic adaptive decisions to scale up based on the sqs queue stats, it returns
 // immediately with a boolean stop channel (sending an event to the channel stops execution).
-func scalingDecisions(sqsClient sqsiface.SQSAPI, lambdaClient lambdaiface.LambdaAPI) chan bool {
-	stopScaling := make(chan bool)
-
+func scalingDecisions(ctx context.Context, sqsClient sqsiface.SQSAPI, lambdaClient lambdaiface.LambdaAPI) {
 	go func() {
 		ticker := time.NewTicker(processingScaleDecisionInterval)
 		defer ticker.Stop()
-		poll := true
-		for poll {
+		pull := true
+		for pull {
 			select {
 			case <-ticker.C:
-			case <-stopScaling:
-				poll = false
+			case <-ctx.Done():
+				pull = false
 			}
 
 			// check if we need to scale
-			totalQueuedMessages, err := queueDepth(sqsClient) // this includes queued and delayed messages
+			totalQueuedMessages, err := queueDepth(ctx, sqsClient) // this includes queued and delayed messages
 			if err != nil {
 				zap.L().Warn("rescale cannot read from sqs queue", zap.Error(err))
 				continue
 			}
 
 			// the number of lambdas to invoke are proportional to the message count (clipped to processingMaxLambdaInvoke)
-			processingScaleUp(lambdaClient, totalQueuedMessages/processingMaxFilesLimit)
+			processingScaleUp(ctx, lambdaClient, totalQueuedMessages/processingMaxFilesLimit)
 		}
 	}()
+}
 
-	return stopScaling
+func queueDepth(ctx context.Context, sqsClient sqsiface.SQSAPI) (int, error) {
+	getQueueAttributesInput := &sqs.GetQueueAttributesInput{
+		AttributeNames: []*string{
+			aws.String(sqs.QueueAttributeNameApproximateNumberOfMessages), // tells us there is waiting events now
+		},
+		QueueUrl: &common.Config.SqsQueueURL,
+	}
+	getQueueAttributesOutput, err := sqsClient.GetQueueAttributesWithContext(ctx, getQueueAttributesInput)
+	if err != nil && err != context.Canceled {
```

In case the context is cancelled, also cancel AWS API call

---

if `err !=nil ` but `err != Cancelled`, do we just ignore the error?

---

Yup.